### PR TITLE
[ROCm] Use AITER sampling implementation instead of torch native

### PR DESCRIPTION
**Pull request by b8zhong**
### Motivation

Note: merge after aiter bump, because of the below issue. There was a fix in https://github.com/ROCm/aiter/pull/1154
> I'm not sure if it's exactly related to gfx950 target. I am in the Docker container.

<details>
<summary>Expand it</summary>

```
*>(indices_ptr), top_k_val, top_p_val, 128256, static_cast<uint64_t>(philox_seed), static_cast<uint64_t>(philox_offset));\n}'}...
[aiter] start build /root/.aiter/build/top_k_top_p_sampling_from_probs_8cf4b5128e35da75bd73cf78ae7dab83
[2025-10-08 21:38:49] start build /root/.aiter/build/top_k_top_p_sampling_from_probs_8cf4b5128e35da75bd73cf78ae7dab83
hipcc -fPIC -mllvm --amdgpu-kernarg-preload-count=16 -D__HIP_PLATFORM_AMD__=1 -fno-offload-uniform-block -Wno-switch-bool -O3 -Wno-undefined-func-template -DUSE_ROCM -std=c++20 -Wno-vla-cxx-extension -mllvm -amdgpu-early-inline-all=true -fgpu-flush-denormals-to-zero -DUSE_PROF_API=1 -Wno-unused-result -mllvm --lsr-drop-solution=1 -mllvm -amdgpu-coerce-illegal-types=1 -DENABLE_FP8 --offload-arch=gfx950 -D__HIP_PLATFORM_HCC__=1 -mllvm -amdgpu-function-calls=false -DLEGACY_HIPBLAS_DIRECT -U__HIP_NO_HALF_CONVERSIONS__ -U__HIP_NO_HALF_OPERATORS__ -mllvm -enable-post-misched=0 -I/root/.aiter/build/top_k_top_p_sampling_from_probs_8cf4b5128e35da75bd73cf78ae7dab83/include -c top_k_top_p_sampling_from_probs_8cf4b5128e35da75bd73cf78ae7dab83.cpp -o top_k_top_p_sampling_from_probs_8cf4b5128e35da75bd73cf78ae7dab83.o
In file included from top_k_top_p_sampling_from_probs_8cf4b5128e35da75bd73cf78ae7dab83.cpp:18:
/root/.aiter/build/top_k_top_p_sampling_from_probs_8cf4b5128e35da75bd73cf78ae7dab83/include/sampling.cuh:299:18: error: reference to non-static member function must be called
  298 |             BlockScan<float, BLOCK_THREADS, SCAN_ALGORITHM>(temp_storage->block_prim.scan)
      |             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  299 |                 .InclusiveSum<VEC_SIZE>(prob_greater_than_threshold, inclusive_cdf);
      |                 ~^~~~~~~~~~~~
/root/.aiter/build/top_k_top_p_sampling_from_probs_8cf4b5128e35da75bd73cf78ae7dab83/include/sampling.cuh:543:13: note: in instantiation of function template specialization 'aiter::sampling::DeviceSamplingFromProb<4U, 1024U, hipcub::BLOCK_SCAN_WARP_SCANS, hipcub::BLOCK_REDUCE_WARP_REDUCTIONS, false, (lambda at /root/.aiter/build/top_k_top_p_sampling_from_probs_8cf4b5128e35da75bd73cf78ae7dab83/include/sampling.cuh:548:23)>' requested here
  543 |             DeviceSamplingFromProb<VEC_SIZE,
      |             ^
top_k_top_p_sampling_from_probs_8cf4b5128e35da75bd73cf78ae7dab83.cpp:45:36: note: in instantiation of function template specialization 'aiter::sampling::TopKTopPSamplingFromProbKernel<1024U, hipcub::BLOCK_SCAN_WARP_SCANS, hipcub::BLOCK_REDUCE_WARP_REDUCTIONS, 4U, false, float, int>' requested here
   45 |     auto kernel = aiter::sampling::TopKTopPSamplingFromProbKernel<aiter::sampling::BLOCK_THREADS, aiter::sampling::SCAN_ALGO, aiter::sampling::REDUCE_ALGO,
      |                                    ^
/opt/rocm-7.0.0/lib/llvm/bin/../../../include/hipcub/block/../backend/rocprim/block/block_scan.hpp:112:10: note: possible target for call
  112 |     void InclusiveSum(T input, T& output)
      |          ^
/opt/rocm-7.0.0/lib/llvm/bin/../../../include/hipcub/block/../backend/rocprim/block/block_scan.hpp:118:10: note: possible target for call
  118 |     void InclusiveSum(T input, T& output, T& block_aggregate)
      |          ^
/opt/rocm-7.0.0/lib/llvm/bin/../../../include/hipcub/block/../backend/rocprim/block/block_scan.hpp:125:10: note: possible target for call
  125 |     void InclusiveSum(T input, T& output, BlockPrefixCallbackOp& block_prefix_callback_op)
      |          ^
/opt/rocm-7.0.0/lib/llvm/bin/../../../include/hipcub/block/../backend/rocprim/block/block_scan.hpp:134:10: note: possible target for call
  134 |     void InclusiveSum(T(&input)[ITEMS_PER_THREAD], T(&output)[ITEMS_PER_THREAD])
      |          ^
/opt/rocm-7.0.0/lib/llvm/bin/../../../include/hipcub/block/../backend/rocprim/block/block_scan.hpp:141:10: note: possible target for call
  141 |     void InclusiveSum(T(&input)[ITEMS_PER_THREAD], T(&output)[ITEMS_PER_THREAD],
      |          ^
/opt/rocm-7.0.0/lib/llvm/bin/../../../include/hipcub/block/../backend/rocprim/block/block_scan.hpp:149:10: note: possible target for call
  149 |     void InclusiveSum(T(&input)[ITEMS_PER_THREAD], T(&output)[ITEMS_PER_THREAD],
      |          ^
1 error generated when compiling for gfx950.
```

</details>

Note: ROCm only change, I tested this on MI355X.

There is a sampling operators available. It's more performant than the torch native implementation. Note this does not affect the greedy case, as if we have top_p = 1 and top_k = 1, we will still use `torch.argmax`. There was another operator for this, but I found that it has some correctness issues (or TBD maybe my fault of the usage), so it was not integrated in this PR.

### Before
**Command**
```bash
lm_eval \
  --model local-completions \
  --tasks gsm8k_platinum \
  --model_args model=amd/Llama-3.1-8B-Instruct-FP8-KV,base_url=http://localhost:30000/v1/completions \
  --trust_remote_code \
  --num_fewshot 8 \
  --batch_size 256 \
  --gen_kwargs "do_sample=True,temperature=0.7,top_p=0.9,top_k=50,max_new_tokens=256"
```

**Results**
```
local-completions (model=amd/Llama-3.1-8B-Instruct-FP8-KV,base_url=http://localhost:30000/v1/completions,trust_remote_code=True), gen_kwargs: (do_sample=True,temperature=0.7,top_p=0.9,top_k=50,max_new_tokens=256), limit: None, num_fewshot: 8, batch_size: 256
|    Tasks     |Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|--------------|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k_platinum|      3|flexible-extract|     8|exact_match|↑  |0.7535|±  |0.0124|
|              |       |strict-match    |     8|exact_match|↑  |0.7304|±  |0.0128|
```

### After
**Command**
```bash
lm_eval \
  --model local-completions \
  --tasks gsm8k_platinum \
  --model_args model=amd/Llama-3.1-8B-Instruct-FP8-KV,base_url=http://localhost:30000/v1/completions \
  --trust_remote_code \
  --num_fewshot 8 \
  --batch_size 256 \
  --gen_kwargs "do_sample=True,temperature=0.7,top_p=0.9,top_k=50,max_new_tokens=256"
```

**Results**
```
local-completions (model=amd/Llama-3.1-8B-Instruct-FP8-KV,base_url=http://localhost:30000/v1/completions,trust_remote_code=True), gen_kwargs: (do_sample=True,temperature=0.7,top_p=0.9,top_k=50,max_new_tokens=256), limit: None, num_fewshot: 8, batch_size: 256
|    Tasks     |Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|--------------|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k_platinum|      3|flexible-extract|     8|exact_match|↑  |0.7593|±  |0.0123|
|              |       |strict-match    |     8|exact_match|↑  |0.7146|±  |0.0130|
```
Generally, the results are inline with the existing sampling behaviour.

---

### Before

**Command**
```bash
python3 -m sglang.bench_serving \
  --backend sglang \
  --host localhost \
  --random-range-ratio 1 \
  --port 30000 \
  --num-prompts 8192 \
  --random-input-len 1024 \
  --random-output-len 1024 \
  --max-concurrency 64 \
  --flush-cache \
  --extra-request-body '{"sampling_params": {"top_p": 0.8, "top_k": 50}}'
```

**Results**
```
============ Serving Benchmark Result ============
Backend:                                 sglang    
Traffic request rate:                    inf       
Max request concurrency:                 64        
Successful requests:                     8192      
Benchmark duration (s):                  184.12    
Total input tokens:                      2477731   
Total generated tokens:                  1029839   
Total generated tokens (retokenized):    1026274   
Request throughput (req/s):              44.49     
Input token throughput (tok/s):          13457.40  
Output token throughput (tok/s):         5593.41   
Total token throughput (tok/s):          19050.81  
Concurrency:                             63.59     
----------------End-to-End Latency----------------
Mean E2E Latency (ms):                   1429.19   
Median E2E Latency (ms):                 1409.63   
---------------Time to First Token----------------
Mean TTFT (ms):                          127.97    
Median TTFT (ms):                        115.19    
P99 TTFT (ms):                           493.79    
---------------Inter-Token Latency----------------
Mean ITL (ms):                           10.46     
Median ITL (ms):                         8.08      
P95 ITL (ms):                            21.46     
P99 ITL (ms):                            42.90     
Max ITL (ms):                            514.47    
==================================================
```
### After

**Command**
```bash
python3 -m sglang.bench_serving \
  --backend sglang \
  --host localhost \
  --random-range-ratio 1 \
  --port 30000 \
  --num-prompts 8192 \
  --random-input-len 1024 \
  --random-output-len 1024 \
  --max-concurrency 64 \
  --flush-cache \
  --extra-request-body '{"sampling_params": {"top_p": 0.8, "top_k": 50}}'
```

**Results**
```
============ Serving Benchmark Result ============
Backend:                                 sglang    
Traffic request rate:                    inf       
Max request concurrency:                 64        
Successful requests:                     8192      
Benchmark duration (s):                  174.41    
Total input tokens:                      2477731   
Total generated tokens:                  1019113   
Total generated tokens (retokenized):    1011151   
Request throughput (req/s):              46.97     
Input token throughput (tok/s):          14206.37  
Output token throughput (tok/s):         5843.21   
Total token throughput (tok/s):          20049.58  
Concurrency:                             63.41     
----------------End-to-End Latency----------------
Mean E2E Latency (ms):                   1350.03   
Median E2E Latency (ms):                 1341.18   
---------------Time to First Token----------------
Mean TTFT (ms):                          147.44    
Median TTFT (ms):                        126.67    
P99 TTFT (ms):                           719.56    
---------------Inter-Token Latency----------------
Mean ITL (ms):                           9.80      
Median ITL (ms):                         7.22      
P95 ITL (ms):                            25.73     
P99 ITL (ms):                            39.95     
Max ITL (ms):                            566.87    
==================================================
```

The changes improved throughput by roughly 6%. After some further testing for various top_p and top_k values in individual sampling, it outperforms by a decent margin in nearly all combinations.

## Summary by CodeRabbit

- New Features
  - Added AITER sampling backend with automatic selection on HIP/ROCm; falls back to existing backends otherwise.
  - New CLI choice for --sampling-backend: aiter.
  - Added an environment flag to control AITER availability.
  - Introduced a public helper to apply custom logit processors to logits.

- Performance
  - Faster greedy and probabilistic sampling when using AITER, with graceful fallback for unsupported options (e.g., min_p).

### CODE REVIEW

File: python/sglang/srt/layers/sampler.py [hunk 1]
```diff
@@ -58,6 +83,69 @@ def _preprocess_logits(
...
+            )
+            return next_token_ids.view(-1)
+        if use_top_k:
+            renorm_probs = ops.top_k_renorm_probs(
+                probs_for_ops,
+                *_to_tensor_scalar_tuple(sampling_info.top_ks),
+            )
+            return torch.multinomial(renorm_probs, num_samples=1).view(-1)
+
+        return torch.multinomial(probs_for_ops, num_samples=1).view(-1)
```
[the remark this answers is not in the document]
Comment (b8zhong) [reply python/sglang/srt/layers/sampler.py]: Done

File: python/sglang/srt/layers/sampler.py [hunk 2]
```diff
@@ -58,6 +83,69 @@ def _preprocess_logits(
 
         return logits
 
+    def _aiter_greedy_sample(self, logits: torch.Tensor) -> torch.Tensor:
+        ops = _load_aiter_ops()
+        del ops  # ensure kernels are registered before calling python API
```
[the remark this answers is not in the document]
Comment (b8zhong) [reply python/sglang/srt/layers/sampler.py]: Ok

File: python/sglang/srt/environ.py [hunk 3]
```diff
@@ -156,6 +156,7 @@ class Envs:
 
     # AMD & ROCm
     SGLANG_USE_AITER = EnvBool(False)
+    SGLANG_IS_AITER_AVAILABLE = EnvBool(True)
```
Comment (hubertlu-tw) [inline python/sglang/srt/environ.py]: Do we need this env var?

**Review by hubertlu-tw — CHANGES_REQUESTED** [review]
Could you please add a test for _aiter_sample or point out which current tests cover it?
Maybe add one test called "test_aiter_sampling_backend.py" like
https://github.com/sgl-project/sglang/blob/main/test/srt/test_pytorch_sampling_backend.py

and add it to [run_suite.py's per-commit-amd](https://github.com/sgl-project/sglang/blob/main/test/srt/run_suite.py#L200).

File: python/sglang/srt/environ.py (hunk 3 above)
Comment (b8zhong) [reply python/sglang/srt/environ.py]: Forgot it in refactor. Removed 👍

**Comment by b8zhong** [comment]
Discussed on Slack, but:
> Added UT, but due to https://github.com/ROCm/aiter/issues/1137, only merge after aiter upgrade past [v0.1.6](https://github.com/ROCm/aiter/releases/tag/v0.1.6)
